src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,238 @@
+            out ExpressionSyntax result)
+        {
+            if (s_binaryMap.TryGetValue(expression.Kind(), out var negatedExpressionInfo))
+            {
+                var binaryExpression = (BinaryExpressionSyntax)expression;
+                var (negatedExpressionType, negatedOperatorType) = negatedExpressionInfo;
+
+                // Certain expressions can never be negative, such as length or unsigned numeric types.
+                // If these expressions are compared with zero using <, >, or =, we construct the negated
+                // binary expression to reflect that it will never be negative.
+                // For example, the expression Array.Length > 0, becomes Array.Length == 0 when negated.
+                var operation = semanticModel.GetOperation(binaryExpression);
+
+                if (IsSpecialCaseBinaryExpression(operation as IBinaryOperation, cancellationToken))
+                {
+                    negatedOperatorType = binaryExpression.OperatorToken.Kind() == SyntaxKind.EqualsEqualsToken
+                        ? binaryExpression.Right is LiteralExpressionSyntax ? SyntaxKind.GreaterThanToken : SyntaxKind.LessThanToken
+                        : SyntaxKind.EqualsEqualsToken;
+                    negatedExpressionType = binaryExpression.Kind() == SyntaxKind.EqualsExpression
+                        ? binaryExpression.Right is LiteralExpressionSyntax ? SyntaxKind.GreaterThanExpression : SyntaxKind.LessThanExpression
+                        : SyntaxKind.EqualsExpression;
+                }
+
+                result = SyntaxFactory.BinaryExpression(
+                    negatedExpressionType,
+                    binaryExpression.Left,
+                    SyntaxFactory.Token(
+                        binaryExpression.OperatorToken.LeadingTrivia,
+                        negatedOperatorType,
+                        binaryExpression.OperatorToken.TrailingTrivia),
+                    binaryExpression.Right);
+
+                return true;
+            }
+
+            result = null;
+            return false;
+        }
+
+        private ExpressionSyntax Negate(ExpressionSyntax expression, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            if (TryNegateBinaryComparisonExpression(expression, semanticModel, cancellationToken, out var result))
+            {
+                return result;
+            }
+
+            switch (expression.Kind())
+            {
+                case SyntaxKind.ParenthesizedExpression:
+                {
+                    var parenthesizedExpression = (ParenthesizedExpressionSyntax)expression;
+                    return parenthesizedExpression
+                        .WithExpression(Negate(parenthesizedExpression.Expression, semanticModel, cancellationToken))
+                        .WithAdditionalAnnotations(Simplifier.Annotation);
+                }
+
+                case SyntaxKind.LogicalNotExpression:
+                {
```

Comment:
nit: consider breaking out all these complex cases into helper methods you dispatch to.